util/packages.go

```diff
@@ -19,84 +20,326 @@ import (
 // PackageValidationDisabled is a flag which can disable package content validation (package, data streams, assets, etc.).
 var PackageValidationDisabled bool
 
-var packageList Packages
+// Packages is a list of packages.
+type Packages []*Package
 
-type Packages []Package
+func (p Packages) Len() int      { return len(p) }
+func (p Packages) Swap(i, j int) { p[i], p[j] = p[j], p[i] }
+func (p Packages) Less(i, j int) bool {
+	if p[i].Name == p[j].Name {
+		return p[i].Version < p[j].Version
+	}
+	return p[i].Name < p[j].Name
+}
+
+// Join returns a set of packages that combines both sets.
+func (p1 Packages) Join(p2 Packages) Packages {
+	// TODO: Avoid duplications?
+	return append(p1, p2...)
+}
+
+// GetPackagesOptions can be used to pass options to GetPackages.
+type GetPackagesOptions struct {
+	// Filter to apply when querying for packages. If the filter is nil,
+	// all packages are returned. This is different to a zero-object filter,
+	// where internal and experimental packages are filtered by default.
+	Filter *PackageFilter
+}
+
+// FileSystemIndexer indexes packages from the filesystem.
+type FileSystemIndexer struct {
+	paths       []string
+	packageList Packages
+
+	// Label used for APM instrumentation.
+	label string
+
+	// Walker function used to look for files.
+	walkerFn func(basePath, path string, info os.FileInfo, err error) error
+
+	// Builder to access the files of a package in this indexer.
+	fsBuilder FileSystemBuilder
+}
+
+var walkerIndexFile = errors.New("file should be indexed")
+
+// NewFileSystemIndexer creates a new FileSystemIndexer for the given paths.
+func NewFileSystemIndexer(paths ...string) *FileSystemIndexer {
+	walkerFn := func(basePath, path string, info os.FileInfo, err error) error {
+		relativePath, err := filepath.Rel(basePath, path)
+		if err != nil {
+			return err
+		}
+
+		dirs := strings.Split(relativePath, string(filepath.Separator))
+		if len(dirs) < 2 {
+			return nil // need to go to the package version level
+		}
+
+		if info.IsDir() {
+			versionDir := dirs[1]
+			_, err := semver.StrictNewVersion(versionDir)
+			if err != nil {
+				log.Printf("warning: unexpected directory: %s, ignoring", path)
+				return filepath.SkipDir
+			}
+			return walkerIndexFile
+		}
+		// Unexpected file, return nil in order to continue processing sibling directories
+		// Fixes an annoying problem when the .DS_Store file is left behind and the package
+		// is not loading without any error information
+		log.Printf("warning: unexpected file: %s, ignoring", path)
+		return nil
+	}
+	fsBuilder := func(p *Package) (PackageFileSystem, error) {
+		return NewExtractedPackageFileSystem(p)
+	}
+	return &FileSystemIndexer{
+		paths:     paths,
+		label:     "FileSystemIndexer",
+		walkerFn:  walkerFn,
+		fsBuilder: fsBuilder,
```

Comment:
maybe just:

```
fsBuilder: NewExtractedPackageFileSystem,
```

?

Reply:
`NewExtractedPackageFileSystem` returns the specific type `*ExtractedPackageFileSystem`, fsBuilder expects a builder that returns the `FileSystemBuilder` interface.

Reply:
Ok, I see now. My intention was to get rid of the additional proxy function due to lang limitations. If it isn't possible, feel free to ignore this comment.